python/ray/train/ml_trainer.py

```diff
@@ -0,0 +1,108 @@
+import abc
+import logging
+from typing import Dict, Union, Callable, Optional
+
+from ray.data import Dataset
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.result import Result
+from ray import tune
+from ray.tune.trainable import ConvertibleToTrainable
+from ray.tune.utils import deep_update
+
+GenDataset = Union[Dataset, Callable[[], Dataset]]
+
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: support conditional hyperparameter tuning of Trainer __init__ args.
+class Trainer(ConvertibleToTrainable, abc.ABC):
+    """
+    Defines interface for distributed training on Ray.
+
+    Args:
+        scaling_config (Optional[ScalingConfig]): Configuration for how to
+            scale training.
+        run_config (Optional[RunConfig]): Configuration for the execution of
+            the training run.
+        train_dataset (Optional[GenDataset]): Either a distributed Ray
+            :ref:`Dataset <dataset-api>` or a Callbable that returns a Dataset,
+            to use for training. If a ``preprocessor`` is also provided,
+            it will be fit on this dataset and this dataset will be
+            transformed.
+        additional_datasets (Optional[GenDataset]): Any additional
+            Datasets (such as validation or test datasets) to use for
+            training. If a ``preprocessor`` is provided, the datasets
+            specified here will only be transformed, and not fit on.
+        preprocessor (Optional[Preprocessor]): A preprocessor to preprocess
+            the provided datasets.
+        resume_from_checkpoint (Optional[Checkpoint]): A checkpoint to
+            resume training from.
+    """
+
+    def __init__(
+        self,
+        scaling_config: Optional[ScalingConfig] = None,
+        run_config: Optional[RunConfig] = None,
+        train_dataset: Optional[GenDataset] = None,
+        additional_datasets: Optional[Dict[str, GenDataset]] = None,
+        preprocessor: Optional[Preprocessor] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+        self.scaling_config = scaling_config
+        self.run_config = run_config if run_config else {}
+        self.train_dataset = train_dataset
+        self.additional_datasets = additional_datasets
+        self.preprocessor = preprocessor
+        self.checkpoint_to_resume_from = resume_from_checkpoint
+
+    def fit(self) -> Result:
+        """Runs training.
+
+        Returns:
+            A Result object containing the training result.
+        """
+        trainable = self.as_trainable()
+
+        # Copied from initial prototyping.
+        # TODO: Replace with Tuner.
```

Comment:
Please add owner for TODOs, e.g., TODO(amog)

Reply:
Added!